File: Codes/1472-design-browser-history.java
/*
 * @Author: LetMeFly
 * @Date: 2025-02-26 13:41:42
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-02-26 13:45:07
 */
import java.util.List;
import java.util.ArrayList;

class BrowserHistory {
    private List<String> history;
    private int now, right;

    public BrowserHistory(String homepage) {
        history = new ArrayList<>();
        now = right = 0;
        history.add(homepage);
    }
    
    public void visit(String url) {
        if (++now == history.size()) {
            history.add(url);
        } else {
            history.set(now, url);
        }
        right = now;
    }
    
    public String back(int steps) {
        now = Math.max(0, now - steps);
        return history.get(now);
    }
    
    public String forward(int steps) {
        now = Math.min(right, now + steps);
        return history.get(now);
    }
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory obj = new BrowserHistory(homepage);
 * obj.visit(url);
 * String param_2 = obj.back(steps);
 * String param_3 = obj.forward(steps);
 */

File: Codes/1472-design-browser-history.go
/*
 * @Author: LetMeFly
 * @Date: 2025-02-26 13:45:43
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-02-26 13:49:00
 */
package main

type BrowserHistory struct {
    history []string
    now,
    right int
}


func Constructor(homepage string) BrowserHistory {
    history := make([]string, 1)
    history[0] = homepage
    return BrowserHistory{
        history: history,
        now: 0,
        right: 0,
    }
}


func (this *BrowserHistory) Visit(url string)  {
    this.now++
    if this.now == len(this.history) {
        this.history = append(this.history, url)
    } else {
        this.history[this.now] = url
    }
    this.right = this.now
}


func (this *BrowserHistory) Back(steps int) string {
    this.now = max(0, this.now - steps)
    return this.history[this.now]
}


func (this *BrowserHistory) Forward(steps int) string {
    this.now = min(this.right, this.now + steps)
    return this.history[this.now]
}


/**
 * Your BrowserHistory object will be instantiated and called as such:
 * obj := Constructor(homepage);
 * obj.Visit(url);
 * param_2 := obj.Back(steps);
 * param_3 := obj.Forward(steps);
 */